### PR TITLE
VPN-4921 - Get all iOS logs to show up on console and log file

### PR DESCRIPTION
**Pull request by brizental**
I built this on top of #7256. Will rebase once that merged.

This PR starts at 59aa9126e12157fb9b7b4899c066007196b60e71.

### CODE REVIEW

File: src/apps/vpn/platforms/ios/ioslogger.mm [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  if (impl) {
+    [impl dealloc];
+    impl = nullptr;
+  }
+}
+
+// static
+void IOSLogger::debug(const QString& message) {
+    [impl debugWithMessage:message.toNSString()];
+}
```
Comment (brizental) [inline src/apps/vpn/platforms/ios/ioslogger.mm:37]: nit: Make formatting consistent.

File: src/apps/vpn/platforms/ios/ioscontroller.mm [hunk 2]
```diff
@@ -217,29 +217,11 @@ emit statusUpdated(QString::fromNSString(serverIpv4Gateway),
...
-  }
-
-  QByteArray content = file.readAll();
-  callback(content);
+  [IOSLoggerImpl getAppexLogsWithCallback:^(NSString* logs) {
+    callback(QString::fromNSString(logs));
+  }];
 }
 
 void IOSController::cleanupBackendLogs() {
```
Comment (brizental) [inline src/apps/vpn/platforms/ios/ioscontroller.mm:225]: I noticed this was not getting called ever. I am not sure if that bug is related here, i should check it before merging. I think it's most likely a separate bug. I'll get back here with an answer.

**Review by mcleinman — COMMENTED** [review]
Couple small things. This is looking great, thank you.

File: src/apps/vpn/platforms/ios/ioslogger.swift [hunk 3]
```diff
@@ -4,49 +4,108 @@
...
+    @objc static func clearAppexLogs() {
+        withAppexLogFile { logFileHandle in
+            logFileHandle.truncateFile(atOffset: 0)
+        }
+    }
+    
+    private static func withAppexLogFile(_ f: (_ handle: FileHandle) throws -> Void) {
+        if (IOSLoggerImpl.appexLogFileURL == nil) {
+            logger.error(message: "IMPOSSIBLE: No known app extension log file.")
+        }
```
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/platforms/ios/ioslogger.swift:79]: Fancy, I like it 💅

File: src/apps/vpn/platforms/ios/ioslogger.swift [hunk 4]
```diff
@@ -4,49 +4,108 @@
...
-        Logger.global?.log(message: "App version: \(appVersion)")
     }
 }
 
-func wg_log(_ type: OSLogType, staticMessage msg: StaticString) {
-    os_log(msg, log: OSLog.default, type: type)
-    Logger.global?.log(message: "\(msg)")
+// The following functions are used by Wireguard internally for logging.
+
+let WIREGUARD_LOGGER = IOSLoggerImpl(tag: "Wireguard")
```
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/platforms/ios/ioslogger.swift:102]: The all caps is usual for global variables, which this one is. Is that convention different in Swift?

File: src/apps/vpn/platforms/ios/iostunnelmanager.swift [hunk 5]
```diff
@@ -56,40 +62,42 @@ class TunnelManager {
...
             if self == nil {
-                Logger.global?.log(message: "We are shutting down.")
+                TunnelManager.logger.debug(message: "We are shutting down.")
+                completionHandler();
                 return
             }
 
             if let error = error {
-                Logger.global?.log(message: "Loading from preference failed: \(error)")
+                TunnelManager.logger.error(message: "Loading from preference failed: \(error)")
```
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/platforms/ios/iostunnelmanager.swift:83]: `logger` is a static variable and this is not a static function.

[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/platforms/ios/iostunnelmanager.swift:83]: 🤦🏻  Thanks

File: src/apps/vpn/platforms/ios/ioslogger.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/platforms/ios/ioslogger.swift:104]: Yes, it differs in Swift. From [Swift guidelines](https://www.swift.org/documentation/api-design-guidelines/):
`Names of types and protocols are UpperCamelCase. Everything else is lowerCamelCase.`

That said, I sometimes see globals/constants in upper camel case, but I've adjusted my suggestion above to lower camel case.

File: src/apps/vpn/platforms/ios/ioslogger.swift [hunk 6]
```diff
@@ -4,49 +4,110 @@
...
 import os.log
+import OSLog
 
-public class Logger {
-    static var global: Logger?
-
-    var tag: String
+public class IOSLoggerImpl : NSObject {
+    private let log: OSLog
+    private let dateFormatter = DateFormatter()
```
Comment (mcleinman) [inline src/apps/vpn/platforms/ios/ioslogger.swift:11]: Nit: If you want to move the formatter setup out of the function, you could get fancy:
```suggestion
    private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd.MM.yyyy HH:mm:ss"   
        return formatter
    }() = DateFormatter()
```

The snippet of setup code will be run the first time the `dateFormatter` is called, and never again. Unfortunately, as of now it needs to be done with a `var` and not `let`.

File: src/apps/vpn/platforms/ios/ioscontroller.mm (hunk 2 above)
Comment (brizental) [reply src/apps/vpn/platforms/ios/ioscontroller.mm:225]: This is an unrelated bug. Filed https://mozilla-hub.atlassian.net/browse/VPN-5101

File: src/apps/vpn/platforms/ios/ioslogger.swift [hunk 7]
```diff
@@ -4,49 +4,114 @@
...
     }
 
-    static func configureGlobal(tagged tag: String, withFilePath filePath: String?) {
-        if Logger.global != nil {
+    func log(_ message: String, type: OSLogType) {
+        os_log("%{public}@", log: self.log, type: type, message)
+        
+        if (Bundle.main.bundlePath.hasSuffix(".appex")) {
+            let currentDate = Date()
+            dateFormatter.dateFormat = "dd.MM.yyyy HH:mm:ss"
```
Comment (mcleinman) [inline src/apps/vpn/platforms/ios/ioslogger.swift:54]:
```suggestion
```

With that new initializer, I believe we can delete this line.

File: src/apps/vpn/platforms/ios/ioslogger.swift [hunk 8]
```diff
@@ -4,49 +4,113 @@
...
-        if let appBuild = Bundle.main.infoDictionary?["CFBundleVersion"] as? String {
-            appVersion += " (\(appBuild))"
+        
+        do {
+            if !FileManager.default.fileExists(atPath: appexLogFileURL.path) {
+                // Create an empty file
+                if let data = "".data(using: .utf8) {
+                    try data.write(to: appexLogFileURL)
+                } else {
+                    logger.error(message: "Unable to create log file at \(appexLogFileURL)")
```
Comment (brizental) [inline src/apps/vpn/platforms/ios/ioslogger.swift:92]:
```suggestion
                    logger.error(message: "Unable to create log file at \(appexLogFileURL)")
                    return
```

**Review by mcleinman — CHANGES_REQUESTED** [review]
Code looks good, but it's not compiling for me or on TaskCluster. I think there are a couple things that were updated in one place but not another (causing an build issue), and a syntax problem. Let me know if you want any help with it.

**Comment by brizental** [comment]
>  Code looks good, but it's not compiling for me or on TaskCluster. I think there are a couple things that were updated in one place but not another (causing an build issue), and a syntax problem. Let me know if you want any help with it.

Sorry for the bitrot here. This one kinda fell through the cracks 😳 Thanks for the patience, it should be working now!

**Comment by flodolo** [comment]
Not sure what't happening with the addons being removed, but I assume this doesn't actually have strings to review?

**Comment by brizental** [comment]
> Not sure what't happening with the addons being removed, but I assume this doesn't actually have strings to review?

Yeah I am so sorry. I made a little bit of a mess with the git history. Please ignore, trying to fix it now 👀

**Review by mcleinman — APPROVED** [review]
This is awesome, thank you! Tests great!

**Pull request merged**